### DISCUSSION TITLE
0.23.0-rc2: Go -> JSVM middleware not working for Handler Func when defined in Go

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#5614] [Q&A] [answered]
I can't get a middleware defined in Golang to work in JS when using handler functions.

$restrict1 doesn't work in gojo land.
```golang
		vm.Set("$restrict1", func(e *core.RequestEvent) error {
			fmt.Println("Restrict Middleware handler func")

			return e.Next()
		})

		vm.Set("$restrict2", &hook.Handler[*core.RequestEvent]{
			Func: func(e *core.RequestEvent) error {
				fmt.Println("Restrict Middleware handler")

				return e.Next()
			},
		})
```

```js
routerAdd("get", "/hello1", (e) => {  
  return e.html(200, "Hello1")
}, $restrict1)

routerAdd("get", "/hello2", (e) => {  
  return e.html(200, "Hello2")
}, $restrict2)
```

I'm getting ` - [routerAdd] failed to wrap middlewares: unsupported goja middleware type`

**Comment by ganigeorgiev** [comment]
Hm, this is because in the type assertion switch we use the [`hook.HandlerFunc[*core.RequestEvent]`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.23.0-rc2/tools/hook#HandlerFunc) type definition and not the the `func(e *core.RequestEvent) error` type.
I'll add it with the final release.

**Reply by dschissler** [reply]
I'll check it if you push it to `develop` now.

**Reply by ganigeorgiev** [reply]
If you don't want to wait you can workaround it with something like:

```go
var restrict1 hook.HandlerFunc[*core.RequestEvent] = func(e *core.RequestEvent) error {
        println("Restrict Middleware handler func")

        return e.Next()
}

...

vm.Set("$restrict1", restrict1)
```

**Reply by dschissler** [reply]
I'm just going to make all of my package middleware give up `hook.Handler` like how PB apis does it.

**Reply by ganigeorgiev** [reply]
There are more or less the same if you are not going to specify an `Id` or `Priority`.

Eventually if https://github.com/golang/go/issues/12854 is implemented (essentially "named arguments") `HandlerFunc` could be soft-deprecated and have only 1 supported bind, aka. `Bind({Func: ... })`

**Reply by dschissler** [reply]
@ganigeorgiev I'll say that I'm a fan of how you removed one layer of function nesting by using `e.Next()`.  I have echo middleware setup functions that are three levels deep, and then wrapped in another level for the builder object.  It's pretty nice to be able to knock that down one layer.  It's vastly easier to read now.

**Reply by dschissler** [reply]
@ganigeorgiev This is extra weird because the gojo main route handler doesn't accept `hook.Handler` at all, and the middleware parameters don't accept `hook.HandlerFunc` at all.

**Reply by ganigeorgiev** [reply]
I'm confused and not sure that I understand what you are trying to do. Please be more specific.

**Reply by ganigeorgiev** [reply]
I'm still not sure that I understand what is the issue but if the `HandlerFunc` type is causing too much confusion we can eventually remove it. Technically they should be the same and you should be able to cast the func to the type with `HandlerFunc[*core.RequestEvent](func)`.

But again, please elaborate to try to understand the issue.

**Reply by ganigeorgiev** [reply, marked as answer]
I pushed a small change in the [`develop` branch](https://github.com/pocketbase/pocketbase/commit/393b461ea27b5df3d33e58c4b438469d79d06275).

To avoid the confusion and unnecessary casting I've decided to remove the `hook.HandlerFunc[T]` type and instead use directly the underlying func definition (`func(T) error`; _outside of the hook package it is actually shorter to write_).

In any case, the route action must be always a function (this is the same with the Go APIs) because it can't have `Priority` or `Id` and under the hood it is added as "finalizer" in the `Hook.Trigger(event, finalizers...)` method.
